File: rebbystore_Frontend/src/services/purchaseService.ts
import { apiClient } from './apiClient';
import type {
  Purchase,
  PurchaseItem,
  PurchaseStatus,
} from '../types/purchase';

interface BackendPurchase {
  id: number;
  supplierId: number;
  supplierName: string;
  purchaseNumber: string;
  status: PurchaseStatus;
  purchaseDate: string;
  totalCost: number;
  notes: string | null;
  createdAt: string;
  updatedAt: string;
}

interface BackendPurchaseItem {
  id: number;
  productId: number;
  productName: string;
  sku: string;
  quantity: number;
  unitCost: number;
  subtotal: number;
}

interface CreatePurchaseInput {
  supplierId: string;
  purchaseDate: string;
  notes?: string;
}

interface AddPurchaseItemInput {
  productId: string;
  quantity: number;
  unitCost: number;
}

interface CreateAndReceivePurchaseInput {
  supplierId: string;
  purchaseDate: string;
  notes?: string;
  items: AddPurchaseItemInput[];
}

function mapPurchase(
  purchase: BackendPurchase,
  items: PurchaseItem[] = [],
): Purchase {
  return {
    id: String(purchase.id),
    supplierId: String(purchase.supplierId),
    supplier: purchase.supplierName,
    purchaseNumber: purchase.purchaseNumber,
    items,
    totalCost: Number(purchase.totalCost),
    date: purchase.purchaseDate,
    notes: purchase.notes ?? undefined,
    status: purchase.status,
  };
}

function mapPurchaseItem(
  item: BackendPurchaseItem,
): PurchaseItem {
  return {
    id: String(item.id),
    productId: String(item.productId),
    productName: item.productName,
    sku: item.sku,
    quantity: item.quantity,
    costPerUnit: Number(item.unitCost),
    totalCost: Number(item.subtotal),
  };
}

async function getItems(purchaseId: string): Promise<PurchaseItem[]> {
  const response = await apiClient<BackendPurchaseItem[]>(
    `/api/purchases/${Number(purchaseId)}/items`,
  );

  return response.map(mapPurchaseItem);
}

async function getPurchasesByStatus(
  status: PurchaseStatus,
): Promise<BackendPurchase[]> {
  return apiClient<BackendPurchase[]>(
    `/api/purchases/status/${status}`,
  );
}

export const purchaseService = {
  async getAll(): Promise<Purchase[]> {
    const statuses: PurchaseStatus[] = [
      'draft',
      'ordered',
      'received',
      'cancelled',
    ];

    const groups = await Promise.all(
      statuses.map((status) => getPurchasesByStatus(status)),
    );

    const purchases = groups.flat();

    const mapped = await Promise.all(
      purchases.map(async (purchase) => {
        const items = await getItems(String(purchase.id));

        return mapPurchase(purchase, items);
      }),
    );

    return mapped.sort(
      (a, b) =>
        new Date(b.date).getTime() -
        new Date(a.date).getTime(),
    );
  },

  async getById(id: string): Promise<Purchase> {
    const purchase = await apiClient<BackendPurchase>(
      `/api/purchases/${Number(id)}`,
    );

    const items = await getItems(id);

    return mapPurchase(purchase, items);
  },

  async create(
    data: CreatePurchaseInput,
  ): Promise<Purchase> {
    const response = await apiClient<BackendPurchase>(
      '/api/purchases',
      {
        method: 'POST',
        body: JSON.stringify({
          supplierId: Number(data.supplierId),
          purchaseDate: data.purchaseDate,
          notes: data.notes || null,
        }),
      },
    );

    return mapPurchase(response);
  },

  async addItem(
    purchaseId: string,
    data: AddPurchaseItemInput,
  ): Promise<PurchaseItem> {
    const response = await apiClient<BackendPurchaseItem>(
      `/api/purchases/${Number(purchaseId)}/items`,
      {
        method: 'POST',
        body: JSON.stringify({
          productId: Number(data.productId),
          quantity: data.quantity,
          unitCost: data.unitCost,
        }),
      },
    );

    return mapPurchaseItem(response);
  },

  async order(id: string): Promise<Purchase> {
    const response = await apiClient<BackendPurchase>(
      `/api/purchases/${Number(id)}/order`,
      {
        method: 'POST',
      },
    );

    const items = await getItems(id);

    return mapPurchase(response, items);
  },

  async receive(id: string): Promise<Purchase> {
    const response = await apiClient<BackendPurchase>(
      `/api/purchases/${Number(id)}/receive`,
      {
        method: 'POST',
      },
    );

    const items = await getItems(id);

    return mapPurchase(response, items);
  },

  async cancel(id: string): Promise<Purchase> {
    const response = await apiClient<BackendPurchase>(
      `/api/purchases/${Number(id)}/cancel`,
      {
        method: 'POST',
      },
    );

    const items = await getItems(id);

    return mapPurchase(response, items);
  },

  async createAndReceive(
    data: CreateAndReceivePurchaseInput,
  ): Promise<Purchase> {
    const purchase = await this.create({
      supplierId: data.supplierId,
      purchaseDate: data.purchaseDate,
      notes: data.notes,
    });

    for (const item of data.items) {
      await this.addItem(purchase.id, item);
    }

    await this.order(purchase.id);

    return this.receive(purchase.id);
  },

  async getTotalThisMonth(): Promise<number> {
    const purchases = await this.getAll();

    const now = new Date();

    const monthStart = new Date(
      now.getFullYear(),
      now.getMonth(),
      1,
    );

    return purchases
      .filter(
        (purchase) =>
          purchase.status === 'received' &&
          new Date(purchase.date) >= monthStart,
      )
      .reduce(
        (sum, purchase) => sum + purchase.totalCost,
        0,
      );
  },
};

File: rebbystore-backend/src/main/java/com/david/rebbystorebackend/service/PurchaseService.java
package com.david.rebbystorebackend.service;

import com.david.rebbystorebackend.domain.entity.Product;
import com.david.rebbystorebackend.domain.entity.Purchase;
import com.david.rebbystorebackend.domain.entity.PurchaseItem;
import com.david.rebbystorebackend.domain.entity.Supplier;
import com.david.rebbystorebackend.repository.ProductRepository;
import com.david.rebbystorebackend.repository.PurchaseItemRepository;
import com.david.rebbystorebackend.repository.PurchaseRepository;
import com.david.rebbystorebackend.repository.SupplierRepository;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.math.BigDecimal;
import java.time.LocalDate;
import java.time.OffsetDateTime;
import java.util.List;
import java.util.UUID;

@Service
@Transactional
public class PurchaseService {

    private static final String DRAFT = "draft";
    private static final String ORDERED = "ordered";
    private static final String RECEIVED = "received";
    private static final String CANCELLED = "cancelled";

    private final PurchaseRepository purchaseRepository;
    private final PurchaseItemRepository purchaseItemRepository;
    private final SupplierRepository supplierRepository;
    private final ProductRepository productRepository;
    private final InventoryService inventoryService;

    public PurchaseService(
            PurchaseRepository purchaseRepository,
            PurchaseItemRepository purchaseItemRepository,
            SupplierRepository supplierRepository,
            ProductRepository productRepository,
            InventoryService inventoryService
    ) {
        this.purchaseRepository = purchaseRepository;
        this.purchaseItemRepository = purchaseItemRepository;
        this.supplierRepository = supplierRepository;
        this.productRepository = productRepository;
        this.inventoryService = inventoryService;
    }

    public Purchase create(
            Long supplierId,
            LocalDate purchaseDate,
            String notes
    ) {
        Supplier supplier = supplierRepository.findById(supplierId)
                .orElseThrow(() ->
                        new IllegalArgumentException(
                                "Supplier with id " + supplierId + " not found"
                        )
                );

        if (purchaseDate == null) {
            throw new IllegalArgumentException(
                    "Purchase date must not be null"
            );
        }

        OffsetDateTime now = OffsetDateTime.now();

        Purchase purchase = new Purchase();
        purchase.setSupplier(supplier);
        purchase.setPurchaseNumber(generatePurchaseNumber());
        purchase.setStatus(DRAFT);
        purchase.setPurchaseDate(purchaseDate);
        purchase.setTotalCost(BigDecimal.ZERO);
        purchase.setNotes(normalizeOptional(notes));
        purchase.setCreatedAt(now);
        purchase.setUpdatedAt(now);

        return purchaseRepository.save(purchase);
    }

    public PurchaseItem addItem(
            Long purchaseId,
            Long productId,
            int quantity,
            BigDecimal unitCost
    ) {
        Purchase purchase = getById(purchaseId);
        ensureStatus(purchase, DRAFT);

        validateQuantity(quantity);
        validateUnitCost(unitCost);

        Product product = productRepository.findById(productId)
                .orElseThrow(() ->
                        new IllegalArgumentException(
                                "Product with id " + productId + " not found"
                        )
                );

        PurchaseItem item = new PurchaseItem();

        item.setPurchase(purchase);
        item.setProduct(product);
        item.setQuantity(quantity);
        item.setUnitCost(unitCost);

        BigDecimal subtotal =
                unitCost.multiply(BigDecimal.valueOf(quantity));

        item.setSubtotal(subtotal);

        PurchaseItem savedItem = purchaseItemRepository.save(item);

        recalculateTotal(purchase);

        return savedItem;
    }

    public void removeItem(Long purchaseItemId) {
        PurchaseItem item = purchaseItemRepository.findById(purchaseItemId)
                .orElseThrow(() ->
                        new IllegalArgumentException(
                                "Purchase item with id " +
                                        purchaseItemId +
                                        " not found"
                        )
                );

        Purchase purchase = item.getPurchase();

        ensureStatus(purchase, DRAFT);

        purchaseItemRepository.delete(item);

        recalculateTotal(purchase);
    }

    public Purchase order(Long purchaseId) {
        Purchase purchase = getById(purchaseId);

        ensureStatus(purchase, DRAFT);

        List<PurchaseItem> items =
                purchaseItemRepository.findByPurchaseId(purchaseId);

        if (items.isEmpty()) {
            throw new IllegalArgumentException(
                    "Purchase must contain at least one item"
            );
        }

        recalculateTotal(purchase);

        purchase.setStatus(ORDERED);
        purchase.setUpdatedAt(OffsetDateTime.now());

        return purchaseRepository.save(purchase);
    }

    public Purchase receive(Long purchaseId) {
        Purchase purchase = getById(purchaseId);

        ensureStatus(purchase, ORDERED);

        List<PurchaseItem> items =
                purchaseItemRepository.findByPurchaseId(purchaseId);

        if (items.isEmpty()) {
            throw new IllegalArgumentException(
                    "Purchase must contain at least one item"
            );
        }

        for (PurchaseItem item : items) {
            inventoryService.stockInFromPurchase(
                    item.getProduct().getId(),
                    item.getQuantity(),
                    purchase
            );
        }

        purchase.setStatus(RECEIVED);
        purchase.setUpdatedAt(OffsetDateTime.now());

        return purchaseRepository.save(purchase);
    }

    public Purchase cancel(Long purchaseId) {
        Purchase purchase = getById(purchaseId);

        if (!DRAFT.equals(purchase.getStatus())
                && !ORDERED.equals(purchase.getStatus())) {
            throw new IllegalArgumentException(
                    "Only draft or ordered purchases can be cancelled"
            );
        }

        purchase.setStatus(CANCELLED);
        purchase.setUpdatedAt(OffsetDateTime.now());

        return purchaseRepository.save(purchase);
    }

    @Transactional(readOnly = true)
    public Purchase getById(Long id) {
        return purchaseRepository.findById(id)
                .orElseThrow(() ->
                        new IllegalArgumentException(
                                "Purchase with id " + id + " not found"
                        )
                );
    }

    @Transactional(readOnly = true)
    public Purchase getByPurchaseNumber(String purchaseNumber) {
        if (purchaseNumber == null || purchaseNumber.trim().isEmpty()) {
            throw new IllegalArgumentException(
                    "Purchase number must not be blank"
            );
        }

        return purchaseRepository
                .findByPurchaseNumber(purchaseNumber.trim())
                .orElseThrow(() ->
                        new IllegalArgumentException(
                                "Purchase with number '" +
                                        purchaseNumber.trim() +
                                        "' not found"
                        )
                );
    }

    @Transactional(readOnly = true)
    public List<PurchaseItem> getItems(Long purchaseId) {
        getById(purchaseId);

        return purchaseItemRepository.findByPurchaseId(purchaseId);
    }

    @Transactional(readOnly = true)
    public List<Purchase> getBySupplier(Long supplierId) {
        return purchaseRepository.findBySupplierId(supplierId);
    }

    @Transactional(readOnly = true)
    public List<Purchase> getByStatus(String status) {
        if (status == null || status.trim().isEmpty()) {
            throw new IllegalArgumentException(
                    "Purchase status must not be blank"
            );
        }

        return purchaseRepository.findByStatusOrderByPurchaseDateDesc(
                status.trim().toLowerCase()
        );
    }

    private void recalculateTotal(Purchase purchase) {
        List<PurchaseItem> items =
                purchaseItemRepository.findByPurchaseId(purchase.getId());

        BigDecimal total = items.stream()
                .map(PurchaseItem::getSubtotal)
                .reduce(BigDecimal.ZERO, BigDecimal::add);

        purchase.setTotalCost(total);
        purchase.setUpdatedAt(OffsetDateTime.now());

        purchaseRepository.save(purchase);
    }

    private void ensureStatus(Purchase purchase, String expectedStatus) {
        if (!expectedStatus.equalsIgnoreCase(purchase.getStatus())) {
            throw new IllegalArgumentException(
                    "Purchase with number '" +
                            purchase.getPurchaseNumber() +
                            "' must be in status '" +
                            expectedStatus +
                            "'"
            );
        }
    }

    private void validateQuantity(int quantity) {
        if (quantity <= 0) {
            throw new IllegalArgumentException(
                    "Purchase quantity must be greater than zero"
            );
        }
    }

    private void validateUnitCost(BigDecimal unitCost) {
        if (unitCost == null) {
            throw new IllegalArgumentException(
                    "Unit cost must not be null"
            );
        }

        if (unitCost.compareTo(BigDecimal.ZERO) <= 0) {
            throw new IllegalArgumentException(
                    "Unit cost must be greater than zero"
            );
        }
    }

    private String normalizeOptional(String value) {
        if (value == null || value.trim().isEmpty()) {
            return null;
        }

        return value.trim();
    }

    private String generatePurchaseNumber() {
        String purchaseNumber;

        do {
            purchaseNumber =
                    "PUR-" +
                            UUID.randomUUID()
                                    .toString()
                                    .substring(0, 8)
                                    .toUpperCase();
        } while (
                purchaseRepository.existsByPurchaseNumber(purchaseNumber)
        );

        return purchaseNumber;
    }
}